File: gts/src/main/java/com/muratkagan/gts/entities/Song.java
package com.muratkagan.gts.entities;

import java.time.LocalDate;
import java.time.OffsetDateTime;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import jakarta.persistence.*;

@Entity
@Table(name = "songs")
public class Song {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	@Column(name = "id", columnDefinition = "serial")
	private Integer id;

	@Column(name = "artist_id", nullable = false)
	private Integer artistId;

	@Column(name = "title", nullable = false)
	private String title;

	@Column(name = "subtitle")
	private String subtitle;

	@Column(name = "release_date")
	private LocalDate releaseDate;

	@Column(name = "duration_seconds")
	private Integer durationSeconds;

	@Column(name = "language_code")
	private String languageCode;

	@CreationTimestamp
	@Column(name = "created_at", updatable = false)
	private OffsetDateTime createdAt;

	@UpdateTimestamp
	@Column(name = "updated_at")
	private OffsetDateTime updatedAt;

	// Getters and setters

	public Integer getId() {
		return id;
	}

	public void setId(Integer id) {
		this.id = id;
	}

	public Integer getArtistId() {
		return artistId;
	}

	public void setArtistId(Integer artistId) {
		this.artistId = artistId;
	}

	public String getTitle() {
		return title;
	}

	public void setTitle(String title) {
		this.title = title;
	}

	public String getSubtitle() {
		return subtitle;
	}

	public void setSubtitle(String subtitle) {
		this.subtitle = subtitle;
	}

	public LocalDate getReleaseDate() {
		return releaseDate;
	}

	public void setReleaseDate(LocalDate releaseDate) {
		this.releaseDate = releaseDate;
	}

	public Integer getDurationSeconds() {
		return durationSeconds;
	}

	public void setDurationSeconds(Integer durationSeconds) {
		this.durationSeconds = durationSeconds;
	}

	public String getLanguageCode() {
		return languageCode;
	}

	public void setLanguageCode(String languageCode) {
		this.languageCode = languageCode;
	}

	public OffsetDateTime getCreatedAt() {
		return createdAt;
	}

	public void setCreatedAt(OffsetDateTime createdAt) {
		this.createdAt = createdAt;
	}

	public OffsetDateTime getUpdatedAt() {
		return updatedAt;
	}

	public void setUpdatedAt(OffsetDateTime updatedAt) {
		this.updatedAt = updatedAt;
	}
}

File: gtsui/types/song.ts
export type SongCreateDto = {
  artistId: number;
  albumId: number;
  title: string;
  subtitle?: string | null;
  releaseDate?: string | null;
  durationSeconds?: number | null;
  languageCode?: string | null;
  genreIds?: string[] | null;
  instrumentationIds?: string[] | null;
  moodIds?: string[] | null;
};

export type SongUpdateDto = {
  artistId: number;
  albumId: number;
  title: string;
  subtitle?: string | null;
  releaseDate?: string | null;
  durationSeconds?: number | null;
  languageCode?: string | null;
  genreIds?: string[] | null;
  instrumentationIds?: string[] | null;
  moodIds?: string[] | null;
};

export type SongResponseDto = {
  id: number;
  artistId: number;
  albumId: number;
  title: string;
  subtitle?: string | null;
  releaseDate?: string | null;
  durationSeconds?: number | null;
  languageCode?: string | null;
  genres?: string[] | null;
  instrumentations?: string[] | null;
  moods?: string[] | null;
  createdAt: string;
  updatedAt: string;
};

export type SongListItemDto = {
  id: number;
  artistId: number;
  albumId: number;
  title: string;
  subtitle?: string | null;
  releaseDate?: string | null;
  durationSeconds?: number | null;
  languageCode?: string | null;
  genres?: string[] | null;
  instrumentations?: string[] | null;
  moods: string[] | null;
  createdAt: string;
  updatedAt: string;
};
